File: nfc_proof_of_concept/packages/flutter_hce/lib/app_layer/file_access/serializers/tlv_block_serializer.dart
import '../../field.dart';
import '../fields/tlv_block_fields.dart';

/// Serializer for a complete File Control TLV (Tag-Length-Value) block.
/// This class assembles the individual fields into a single 8-byte structure.
class FileControlTlv extends ApduSerializer {
  late final TlvTag tag;
  final TlvLength tagLength = TlvLength.forFileControl;
  late final FileIdField fileId;
  late final MaxFileSizeField maxFileSize;
  final ReadAccessField readAccess = ReadAccessField.granted;
  late final WriteAccessField writeAccess;

  /// Named constructor for a standard NDEF File Control TLV.
  /// Uses the standard NDEF Tag (0x04) and File ID (0xE104).
  FileControlTlv.ndef({
    required int maxNdefFileSize,
    required bool isNdefWritable,
  }) : super(name: "NDEF File Control TLV") {
    tag = TlvTag.ndef;
    fileId = FileIdField.forNdef;
    maxFileSize = MaxFileSizeField(maxNdefFileSize);
    writeAccess = WriteAccessField.fromByte(isNdefWritable ? 0x00 : 0xFF);
  }

  /// Named constructor for a Proprietary File Control TLV.
  /// Uses the standard Proprietary Tag (0x05) and a custom File ID.
  FileControlTlv.proprietary({
    required int proprietaryFileId,
    required int maxProprietaryFileSize,
    required bool isProprietaryWritable,
  }) : super(name: "Proprietary File Control TLV") {
    tag = TlvTag.proprietary;
    fileId = FileIdField(proprietaryFileId);
    maxFileSize = MaxFileSizeField(maxProprietaryFileSize);
    writeAccess =
        WriteAccessField.fromByte(isProprietaryWritable ? 0x00 : 0xFF);
  }

  @override
  void setFields() {
    fields = [
      tag,
      tagLength,
      fileId,
      maxFileSize,
      readAccess,
      writeAccess,
    ];
  }
}

File: nfc_proof_of_concept/packages/flutter_hce/android/src/main/kotlin/com/viridian/flutter_hce/app_layer/file_access/serializers/TlvBlockSerializer.kt
package com.viridian.flutter_hce.app_layer.file_access.serializers

import com.viridian.flutter_hce.app_layer.ApduSerializer
import com.viridian.flutter_hce.app_layer.file_access.fields.*

/**
 * Serializer for a complete File Control TLV (Tag-Length-Value) block.
 * This class assembles the individual fields into a single 8-byte structure.
 */
class FileControlTlv : ApduSerializer {
    private val tag: TlvTag
    private val tagLength: TlvLength = TlvLength.forFileControl
    private val fileId: FileIdField
    private val maxFileSize: MaxFileSizeField
    private val readAccess: ReadAccessField = ReadAccessField.granted
    private val writeAccess: WriteAccessField

    companion object {
        /**
         * Named constructor for a standard NDEF File Control TLV.
         * Uses the standard NDEF Tag (0x04) and File ID (0xE104).
         */
    fun ndef(maxNdefFileSize: Int, isNdefWritable: Boolean): FileControlTlv {
            return FileControlTlv(
                "NDEF File Control TLV",
                TlvTag.ndef,
                FileIdField.forNdef,
                MaxFileSizeField(maxNdefFileSize),
        WriteAccessField.fromWritable(isNdefWritable)
            )
        }

        /**
         * Named constructor for a Proprietary File Control TLV.
         * Uses the standard Proprietary Tag (0x05) and a custom File ID.
         */
    fun proprietary(
            proprietaryFileId: Int,
            maxProprietaryFileSize: Int,
            isProprietaryWritable: Boolean
        ): FileControlTlv {
            return FileControlTlv(
                "Proprietary File Control TLV",
                TlvTag.proprietary,
                FileIdField(proprietaryFileId),
                MaxFileSizeField(maxProprietaryFileSize),
        WriteAccessField.fromWritable(isProprietaryWritable)
            )
        }
    }

    private constructor(
        name: String,
        tag: TlvTag,
        fileId: FileIdField,
        maxFileSize: MaxFileSizeField,
        writeAccess: WriteAccessField
    ) : super(name) {
        this.tag = tag
        this.fileId = fileId
        this.maxFileSize = maxFileSize
        this.writeAccess = writeAccess

        // Register fields in the defined order
        register(this.tag)
        register(this.tagLength)
        register(this.fileId)
        register(this.maxFileSize)
        register(this.readAccess)
        register(this.writeAccess)
    }
}
